src/ttlv/fast_scan.rs

```diff
@@ -0,0 +1,1009 @@
+/// errors without tracking any details of them.  In some cases, it will
+/// delay errors, returning garbage data instead of [`Err`].  If scanning
+/// with [`FastScanner`] fails, a different scanner needs to be used to
+/// ascertain the position and cause of the error.
+#[derive(Clone)]
+#[must_use = "Call 'FastScanner::finish()' to finish scanning"]
+pub struct FastScanner<'a> {
+    /// The currently remaining input.
+    remaining: &'a [[u8; 8]],
+
+    /// A failure word.
+    ///
+    /// This is non-zero if an error was detected and delayed.
+    failure_word: u64,
+}
+
+impl<'a> FastScanner<'a> {
+    /// Construct a new [`FastScanner`].
+    pub fn new(input: &'a [u8]) -> Result<Self, FastScanError> {
+        let num_blocks = input.len() / 8;
+        if !input.len().is_multiple_of(8) {
+            // The input contains a partial block.
+            return Err(FastScanError::assert());
+        }
+
+        let ptr = input.as_ptr().cast::<[u8; 8]>();
+        let input = unsafe { core::slice::from_raw_parts(ptr, num_blocks) };
+        Ok(Self::from_blocks(input))
+    }
+
+    /// Construct a new [`FastScanner`] from a block slice.
+    pub const fn from_blocks(input: &'a [[u8; 8]]) -> Self {
+        Self {
+            remaining: input,
+            failure_word: 0,
+        }
+    }
+
+    /// The remaining input to the scanner.
+    pub const fn remaining(&self) -> &'a [[u8; 8]] {
+        self.remaining
+    }
+
+    /// Whether the scanner is finished.
+    pub const fn is_empty(&self) -> bool {
+        self.remaining.is_empty()
+    }
+
+    /// Whether an error has been delayed.
+    ///
+    /// This should only be checked if the consistency of the previously
+    /// scanned data is important, e.g. if acting on it.  It can be ignored
+    /// while parsing.
+    pub const fn delayed_error(&self) -> bool {
+        self.failure_word != 0
+    }
+
+    /// Whether a whole element is available.
+    ///
+    /// The length of the next element will be checked; if it is entirely loaded
+    /// in the buffer, `true` is returned.  Otherwise, or if there is no next
+    /// element, `false` is returned.
+    pub fn have_next(&self) -> bool {
+        let [head, rest @ ..] = self.remaining else {
+            return false;
+        };
+
+        let length = u32::from_be_bytes(head[4..8].try_into().unwrap());
+        let available = rest.as_flattened().len();
+
+        if usize::MAX as u32 == u32::MAX {
+            // 'usize' is at least as big as 'u32'.
+            available >= length as usize
+        } else {
+            // 'u32' is bigger than 'usize'.
+            available as u32 >= length
+        }
+    }
+
+    /// Finish the scanner.
+    ///
+    /// If there is remaining input, or an error was delayed, [`Err`] is
+    /// returned.  If additional remaining input is allowed, use
+    /// [`Self::finish_non_exhaustive()`] instead.
+    pub fn finish(self) -> Result<(), FastScanError> {
```

Comment:
The caller cannot tell if the call failed because more input is available or because there was an error.

Perhaps if they knew there was more input available they would try and scan more, e.g. another message? Or should they call `have_next()` if they want that ability?

Also, you say in the `must_use()` at the top that `finish()` has to be called, but removing the call to `finish()` doesn't break anything that I can see.

And calling `finish()` isn't even needed if the buffer contains a complete KMIP message and one uses the macro defined `fast_scan()` function to deserialize an entire `RequestMessage`.

I find this API a bit confusing.